lightning/src/util/persist.rs

```diff
@@ -22,15 +25,57 @@ use crate::chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate};
-/// Implementing `KVStorePersister` provides auto-implementations for [`Persister`]
-/// and [`Persist`] traits.  It uses "manager", "network_graph",
-/// and "monitors/{funding_txo_id}_{funding_txo_index}" for keys.
-pub trait KVStorePersister {
-	/// Persist the given writeable using the provided key
-	fn persist<W: Writeable>(&self, key: &str, object: &W) -> io::Result<()>;
+use crate::util::logger::Logger;
+use crate::util::ser::{ReadableArgs, Writeable};
+
+/// The namespace under which the [`ChannelManager`] will be persisted.
+pub const CHANNEL_MANAGER_PERSISTENCE_NAMESPACE: &str = "";
+/// The key under which the [`ChannelManager`] will be persisted.
+pub const CHANNEL_MANAGER_PERSISTENCE_KEY: &str = "manager";
+
+/// The namespace under which [`ChannelMonitor`]s will be persisted.
+pub const CHANNEL_MONITOR_PERSISTENCE_NAMESPACE: &str = "monitors";
+
+/// The namespace under which the [`NetworkGraph`] will be persisted.
+pub const NETWORK_GRAPH_PERSISTENCE_NAMESPACE: &str = "";
+/// The key under which the [`NetworkGraph`] will be persisted.
+pub const NETWORK_GRAPH_PERSISTENCE_KEY: &str = "network_graph";
+
+/// The namespace under which the [`WriteableScore`] will be persisted.
+pub const SCORER_PERSISTENCE_NAMESPACE: &str = "";
+/// The key under which the [`WriteableScore`] will be persisted.
+pub const SCORER_PERSISTENCE_KEY: &str = "scorer";
+
+/// Provides an interface that allows to store and retrieve persisted values that are associated
+/// with given keys.
+///
+/// In order to avoid collisions the key space is segmented based on the given `namespace`s.
+/// Implementations of this trait are free to handle them in different ways, as long as
+/// per-namespace key uniqueness is asserted.
+///
+/// Keys and namespaces are required to be valid ASCII strings and the empty namespace (`""`) is
+/// assumed to be valid namespace.
+pub trait KVStore {
+	/// A reader as returned by [`Self::read`].
+	type Reader: io::Read;
+	/// Returns an [`io::Read`] for the given `namespace` and `key` from which [`Readable`]s may be
+	/// read.
+	///
+	/// Returns an [`ErrorKind::NotFound`] if the given `key` could not be found in the given `namespace`.
+	///
+	/// [`Readable`]: crate::util::ser::Readable
+	/// [`ErrorKind::NotFound`]: io::ErrorKind::NotFound
+	fn read(&self, namespace: &str, key: &str) -> io::Result<Self::Reader>;
+	/// Persists the given data under the given `key`.
+	///
+	/// Will create the given `namespace` if not already present in the store.
+	fn write(&self, namespace: &str, key: &str, buf: &[u8]) -> io::Result<()>;
+	/// Removes any data that had previously been persisted under the given `key`.
+	fn remove(&self, namespace: &str, key: &str) -> io::Result<()>;
+	/// Returns a list of keys that are stored under the given `namespace`.
+	///
+	/// Will return an empty list if the `namespace` is unknown.
+	fn list(&self, namespace: &str) -> io::Result<Vec<String>>;
```

Comment:
Expanding on the clarity question -- I think it would be good to say whether the backend should return this `Vec` sorted or not.

If not, I will need to make sure that I sort the listing when doing things like trying to go through updates in order.

If so, then I will not bother with the extra work of sorting a sorted list. 

I think _most_ backends would return this listing sorted, either naturally or through a manual sort (e.g., most filesystems, and most DBs will either naturally store things like updates sorted, or can do it very efficiently). But I will say, not all possible backends will sort.